src/main/scala/org/broadinstitute/dsde/rawls/expressions/SlickExpressionParsing.scala

```diff
@@ -186,91 +192,112 @@ trait SlickExpressionParser extends JavaTokenParsers {
       // the optional entity rec is used for references. Since we know we are not dealing with a reference here
       // as this is the attribute final func, we can pass in None.
       val attributesOption = attributeQuery.unmarshalAttributes(wsIdAndAttributes.map((_, None))).get(context.workspaceContext.workspaceId)
-      attributesOption.map { attributes => Seq(attributes.getOrElse(attributeName, AttributeNull)) }.getOrElse(Seq.empty)
+      val wsExprResult = attributesOption.map { attributes => Seq(attributes.getOrElse(attributeName, AttributeNull)) }.getOrElse(Seq.empty)
+
+      //Return the value of the expression once for each entity we wanted to evaluate this expression against!
+      context.rootEntities.map( rec => (rec.name, wsExprResult) ).toMap
     }
   }
 
   // root func that gets an entity reference off a workspace
   private def workspaceEntityRefRootFunc(entityRefName: String)(context: SlickExpressionContext): PipeType = {
     for {
+      rootEntity <- exprEvalQuery
       workspace <- workspaceQuery.findByIdQuery(context.workspaceContext.workspaceId)
       workspaceAttribute <- workspaceAttributeQuery if workspaceAttribute.workspaceId === workspace.id
       attribute <- attributeQuery if workspaceAttribute.attributeId === attribute.id && attribute.name === entityRefName
       nextEntity <- entityQuery if attribute.valueEntityRef === nextEntity.id
-    } yield nextEntity
+    } yield (rootEntity.name, nextEntity)
   }
 
   // add pipe to an entity referenced by the current entity
   private def entityNameAttributePipeFunc(entityRefName: String)(context: SlickExpressionContext, queryPipeline: PipeType): PipeType = {
-    for {
-      entity <- queryPipeline
+    (for {
+      (rootEntityName, entity) <- queryPipeline
       entityAttribute <- entityAttributeQuery if entity.id === entityAttribute.entityId
       attribute <- attributeQuery if entityAttribute.attributeId === attribute.id && attribute.name === entityRefName
       nextEntity <- entityQuery if attribute.valueEntityRef === nextEntity.id
-    } yield nextEntity
+    } yield (rootEntityName, nextEntity)).sortBy({case (nm, ent) => ent.name })
```

Comment:
This sort needs to happen here in order to maintain consistent ordering across expressions with long, similar chains of entities. For example, the following two expressions:

```
this.foos.bars.baz
this.foos.bars.qux
```

...aren't guaranteed to return the list of foos and bars in the same order during the join, which means the trailing attributes will be completely jumbled. Props to `SlickSimpleExpressionParserTest test("consistent sorting")` for catching this one.